query/functor.hpp

```diff
@@ -447,6 +447,150 @@ struct GetHLLValueFunctor {
   }
 };
 
+// functor to calculate array length
+template <typename I, typename O>
+struct ArrayLengthFunctor {
+  typedef typename thrust::tuple<I, bool> argument_type;
+  typedef typename thrust::tuple<O, bool> result_type;
+
+  __host__ __device__
+  result_type operator()(argument_type arrVal) const {
+    // should never come here
+    O o;
+    return thrust::make_tuple<O, bool>(o, false);
+  }
+};
+
+template <typename I>
+struct ArrayLengthFunctor<I, uint32_t> {
+  typedef typename thrust::tuple<I, bool> argument_type;
+  typedef typename thrust::tuple<uint32_t, bool> result_type;
+
+  __host__ __device__
+  result_type operator()(argument_type arrVal) const {
+    auto valid = thrust::get<1>(arrVal);
+    if (!valid) {
+      return thrust::make_tuple<uint32_t, bool>(0, false);
+    }
+    uint32_t *valP = reinterpret_cast<uint32_t *>(thrust::get<0>(arrVal));
+    if (valP == NULL) {
+      return thrust::make_tuple<uint32_t, bool>(0, true);
+    }
+    return thrust::make_tuple<uint32_t, bool>(*valP, true);
+  }
+};
+
+// functor to get index-th element from array value
+template <typename O, typename I1, typename I2, typename Enabled = void>
+struct ArrayElementAtFunctor {
+  typedef typename thrust::tuple<I1, bool> argument_type_1;
+  typedef typename thrust::tuple<I2, bool> argument_type_2;
+  typedef typename thrust::tuple<O, bool> result_type;
+
+  __host__ __device__
+  result_type operator()(argument_type_1 arg1, argument_type_2 arg2) const {
+    O o;
+    return thrust::make_tuple<O, bool>(o, false);
+  }
+};
+
+template <typename O, typename I>
+struct ArrayElementAtFunctor<O, I, int,
+    typename std::enable_if<
+        std::is_same<typename std::remove_pointer<I>::type, O>::value>::type> {
+  typedef typename thrust::tuple<I, bool> argument_type_1;
+  typedef typename thrust::tuple<int, bool> argument_type_2;
+  typedef typename thrust::tuple<O, bool> result_type;
+
+  __host__ __device__
+  result_type operator()(argument_type_1 arrVal, argument_type_2 indexT) const {
+    auto valid = thrust::get<1>(arrVal);
+    if (!valid) {
+      O v;
+      return thrust::make_tuple<O, bool>(v, false);
+    }
+    uint32_t *lenP = reinterpret_cast<uint32_t *>(thrust::get<0>(arrVal));
+    int index = thrust::get<0>(indexT);
+    if (lenP == NULL || (index >= 0 && *lenP <= index) ||
+       (index < 0 && *lenP < -index)) {
+      O v;
+      return thrust::make_tuple<O, bool>(v, false);
+    }
+    auto valP = reinterpret_cast<I>(
+        reinterpret_cast<uint8_t*>(thrust::get<0>(arrVal)) + 4);
+    int len = static_cast<int>(*lenP);
+    if (index < 0) {
```

Comment:
are we supporting negative index as well? just curious

Reply:
yes, the negative index will be reverse order, -1 is the last